Append the Storm component's task ID to the metric name, too

dschiavu opened the pull request:
Hi,

This pull request appends the Storm component's task ID to the name of the metric sent to Statsd. 

This is needed in cases of Storm component paralellism, because if a Storm component has multiple instances (tasks) running, each task records it own, separate metric, but the current implementation doesn't differentiate the separate tasks metrics, so the metrics get sent to Statsd under the same name, and then are furtherly aggregated by Statsd into a single metric. I think we need to differentiate the different task's metrics, so I've added the task ID to the metric name that gets sent to Statsd.

See Storm's default `LoggingMetricsConsumer`, which correctly records separate per-task metrics into the log file.

Commit:
Append the Storm component's task ID to the metric name, too

Commit:
Emit values to Statsd as gauges instead of counters.

This is a workaround on Storm's imprecise metric emit interval, which
oscillates wildly around the user-defined metric emit interval value.

As Storm could (and does) emit multiple metrics in a single defined time
window, Statsd counts would be aggregated by Statsd (and of course,
duplicated), but gauges don't aggregate, so they're a safe alternative.

It would be best to make the Statsd metric type emitted
user-configurable.

jatrost commented on src/main/java/com/endgame/storm/metrics/statsd/StatsdMetricConsumer.java:
```diff
@@ -139,7 +139,8 @@ public String toString() {
 				.append(taskInfo.srcWorkerPort).append(".")
-				.append(clean(taskInfo.srcComponentId)).append(".");
+				.append(clean(taskInfo.srcComponentId)).append(".")
+				.append(taskInfo.srcTaskId).append(".");
```
I originally did not include the taskID on purpose.  It is because the task ID assignment seems arbitrary (i.e. the same taskID is not assigned to the same worker/port) across topology restarts.  So, to view this data in any sensible view in graphite, you would have to use sumSeries() and a wildcard on the taskID.  To simplify things I just didn't include it.

I would probably be more open to making this a configurable option (e.g. `metrics.statsd.includeTaskID`) so it can be turned on/off, with the default being off.

dschiavu replied:
+1 It would be great to make the task ID configurable! Especially since appending the task ID is what the the default `LoggingMetricsConsumer` does, so you can emit per-task metrics. In my use case, using `statsd` counters and aggregating metrics component-wide is not an option really, because I want to know per-component instance (i.e. task) metrics of my `KafkaTridentSpout` (it works by having a separate instance of this spout assigned to a separate Kafka partition), so the metrics aren't summable across tasks.

jatrost commented on src/main/java/com/endgame/storm/metrics/statsd/StatsdMetricConsumer.java:
```diff
@@ -169,7 +170,7 @@ public String toString() {
 	public void report(String s, int number) {
 		LOG.debug("reporting: {}={}", s, number);
-		statsd.count(s, number);
+		statsd.recordGaugeValue(s, number);
```
I think this gauge setting only makes sense with the taskID included, otherwise it will likely overwrite a value from another task.

dschiavu replied:
Exactly. Could you make this a configurable option, too? Because of the Storm metrics emit timing not behaving exactly as the specified metric emit interval, emitting the metrics to `statsd` as gauge values would solve the problem of `statsd` summing multiple count values in a single time interval, which is what we don't want to (at least in my case, as I keep a Storm metric running count myself).

Commit:
Parametrize the Statsd metric type via the new
"metrics.statsd.metric_type" property. Allowable values: "counter"
(default) or "gauge".

Commit:
Update README.md -- add section on the new configurable Statsd Metric
Type property "metrics.statsd.metric_type"

Commit:
Make the taskId portion of the Statsd metric name configurable via the
"metrics.statsd.include_task_id" property. If set to true (default is
false), the Statsd metric name will include the Storm Task ID.

dschiavu committed:
Update README.md

Add a new section "Statsd Metric Name" with info on the new configuration property "metrics.statsd.include_task_id".

dschiavu commented:
Hi Jason,

I've added more changes to the pull request, specifically:
- I've made the Storm Task ID portion of the Statsd metric name configurable
- I've made the Statsd metric type configurable

Please review the changes. It would be great if you could merge this into your `master` branch.

Regards,

Danijel

dschiavu commented:
I've also modified the `cleanup()` method to remove the "`:`" (colon) character from the metric name, which is present in Trident topologies component names and must not be sent to Statsd (as per the Statsd protocol spec).

dschiavu commented:
Hi,

Any chance getting these changes into your branch?

Thanks,
Danijel
